File: react-native/android/src/main/java/com/ricoh360/thetableclientreactnative/WlanControlCommandV2Service.kt
package com.ricoh360.thetableclientreactnative

import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.WritableMap
import com.ricoh360.thetableclient.BleCharacteristic
import com.ricoh360.thetableclient.service.WlanControlCommandV2
import com.ricoh360.thetableclient.service.data.ConnectedInfo
import com.ricoh360.thetableclient.service.data.ConnectedWifiInfo
import com.ricoh360.thetableclient.service.data.Proxy
import com.ricoh360.thetableclient.service.data.values.NetworkType
import com.ricoh360.thetableclient.service.data.values.WifiSecurityMode
import java.lang.IllegalArgumentException

const val KEY_WIFI_INFO = "wifiInfo"
const val KEY_ETHERNET = "ethernet"
const val KEY_LTE = "lte"
const val KEY_IS_CONNECTED = "isConnected"
const val KEY_IS_INTERNET_ACCESSIBLE = "isInternetAccessible"
const val KEY_SSID_LIST = "ssidList"
const val KEY_SSID_STEALTH = "ssidStealth"
const val KEY_CONNECTION_PRIORITY = "connectionPriority"
const val KEY_IP_ADDRESS = "ipAddress"
const val KEY_SUBNET_MASK = "subnetMask"
const val KEY_DEFAULT_GATEWAY = "defaultGateway"

object WlanControlCommandV2Service {
  suspend fun setNetworkType(id: Int, value: String, promise: Promise) {
    try {
      val device = ThetaBleClientReactNativeModule.deviceList[id]
      device ?: let {
        promise.reject(Exception(ERROR_MESSAGE_DEVICE_NOT_FOUND))
        return
      }
      val service = device.wlanControlCommandV2
      service ?: let {
        promise.reject(Exception(ERROR_MESSAGE_UNSUPPORTED_SERVICE))
        return
      }
      val enumValue = NetworkType.entries.find {
        it.name == value
      }
      if (enumValue == null) {
        promise.reject(Exception("Network type not found. $value"))
      } else {
        service.setNetworkType(enumValue)
        promise.resolve(null)
      }
    } catch (e: Throwable) {
      promise.reject(e)
    }
  }

  fun setNetworkTypeNotify(
    id: Int,
    enable: Boolean,
    promise: Promise,
    sendNotifyEvent: (WritableMap) -> Unit,
  ) {
    try {
      val device = ThetaBleClientReactNativeModule.deviceList[id]
      if (device == null) {
        promise.reject(Exception(ERROR_MESSAGE_DEVICE_NOT_FOUND))
        return
      }
      val service = device.wlanControlCommandV2
      if (service == null) {
        promise.reject(Exception(ERROR_MESSAGE_UNSUPPORTED_SERVICE))
        return
      }

      if (enable) {
        service.setNetworkTypeNotify { value, error ->
          sendNotifyEvent(
            toNotify(
              id,
              BleCharacteristic.WRITE_SET_NETWORK_TYPE,
              value?.let {
                val map = Arguments.createMap()
                map.putString(NetworkType.keyName, it.name)
                map
              },
              error,
            ),
          )
        }
      } else {
        service.setNetworkTypeNotify(null)
      }
      promise.resolve(null)
    } catch (e: Throwable) {
      promise.reject(e)
    }
  }

  suspend fun getConnectedWifiInfo(id: Int, promise: Promise) {
    try {
      val device = ThetaBleClientReactNativeModule.deviceList[id]
      if (device == null) {
        promise.reject(Exception(ERROR_MESSAGE_DEVICE_NOT_FOUND))
        return
      }
      val service = device.wlanControlCommandV2
      if (service == null) {
        promise.reject(Exception(ERROR_MESSAGE_UNSUPPORTED_SERVICE))
        return
      }
      val value = service.getConnectedWifiInfo()
      promise.resolve(fromTheta(value))
    } catch (e: Throwable) {
      promise.reject(e)
    }
  }

  fun setConnectedWifiInfoNotify(
    id: Int,
    enable: Boolean,
    promise: Promise,
    sendNotifyEvent: (WritableMap) -> Unit,
  ) {
    try {
      val device = ThetaBleClientReactNativeModule.deviceList[id]
      if (device == null) {
        promise.reject(Exception(ERROR_MESSAGE_DEVICE_NOT_FOUND))
        return
      }
      val service = device.wlanControlCommandV2
      if (service == null) {
        promise.reject(Exception(ERROR_MESSAGE_UNSUPPORTED_SERVICE))
        return
      }

      if (enable) {
        service.setConnectedWifiInfoNotify { value, error ->
          sendNotifyEvent(
            toNotify(
              id,
              BleCharacteristic.NOTIFICATION_CONNECTED_WIFI_INFO,
              value?.let {
                fromTheta(it)
              },
              error,
            ),
          )
        }
      } else {
        service.setConnectedWifiInfoNotify(null)
      }
      promise.resolve(null)
    } catch (e: Throwable) {
      promise.reject(e)
    }
  }

  suspend fun scanSsidStart(
    id: Int,
    timeout: Int,
    promise: Promise,
    sendNotifyEvent: (WritableMap) -> Unit,
  ) {
    try {
      val device = ThetaBleClientReactNativeModule.deviceList[id]
      device ?: let {
        promise.reject(Exception(ERROR_MESSAGE_DEVICE_NOT_FOUND))
        return
      }
      val service = device.wlanControlCommandV2
      service ?: let {
        promise.reject(Exception(ERROR_MESSAGE_UNSUPPORTED_SERVICE))
        return
      }
      service.scanSsidStart(timeout, object : WlanControlCommandV2.ScanCallback {
        override fun onCompleted(ssidList: List<String>) {
          super.onCompleted(ssidList)
          sendNotifyEvent(
            toNotify(
              id,
              BleCharacteristic.NOTIFICATION_SCANNED_SSID,
              fromTheta(null, ssidList),
              null,
            ),
          )
        }

        override fun onNotify(ssid: String) {
          super.onNotify(ssid)
          sendNotifyEvent(
            toNotify(
              id,
              BleCharacteristic.NOTIFICATION_SCANNED_SSID,
              fromTheta(ssid, null),
              null,
            ),
          )
        }
      })
      promise.resolve(null)
    } catch (e: Throwable) {
      promise.reject(e)
    }
  }

  suspend fun scanSsidStop(id: Int, promise: Promise) {
    try {
      val device = ThetaBleClientReactNativeModule.deviceList[id]
      device ?: let {
        promise.reject(Exception(ERROR_MESSAGE_DEVICE_NOT_FOUND))
        return
      }
      val service = device.wlanControlCommandV2
      service ?: let {
        promise.reject(Exception(ERROR_MESSAGE_UNSUPPORTED_SERVICE))
        return
      }
      service.scanSsidStop()
      promise.resolve(null)
    } catch (e: Throwable) {
      promise.reject(e)
    }
  }

  suspend fun setAccessPointDynamically(id: Int, params: ReadableMap, promise: Promise) {
    try {
      val device = ThetaBleClientReactNativeModule.deviceList[id]
      device ?: let {
        promise.reject(Exception(ERROR_MESSAGE_DEVICE_NOT_FOUND))
        return
      }
      val service = device.wlanControlCommandV2
      service ?: let {
        promise.reject(Exception(ERROR_MESSAGE_UNSUPPORTED_SERVICE))
        return
      }
      val values = toSetAccessPointParams(params)
      service.setAccessPointDynamically(
        ssid = values.ssid,
        ssidStealth = values.ssidStealth,
        security = values.security,
        password = values.password,
        connectionPriority = values.connectionPriority,
        proxy = values.proxy,
      )
      promise.resolve(null)
    } catch (e: Throwable) {
      promise.reject(e)
    }
  }

  suspend fun setAccessPointStatically(id: Int, params: ReadableMap, promise: Promise) {
    try {
      val device = ThetaBleClientReactNativeModule.deviceList[id]
      device ?: let {
        promise.reject(Exception(ERROR_MESSAGE_DEVICE_NOT_FOUND))
        return
      }
      val service = device.wlanControlCommandV2
      service ?: let {
        promise.reject(Exception(ERROR_MESSAGE_UNSUPPORTED_SERVICE))
        return
      }
      val values = toSetAccessPointParams(params)
      service.setAccessPointStatically(
        ssid = values.ssid,
        ssidStealth = values.ssidStealth,
        security = values.security,
        password = values.password,
        connectionPriority = values.connectionPriority,
        ipAddress = values.ipAddress ?: throw IllegalArgumentException(KEY_IP_ADDRESS),
        subnetMask = values.subnetMask ?: throw IllegalArgumentException(KEY_SUBNET_MASK),
        defaultGateway = values.defaultGateway
          ?: throw IllegalArgumentException(KEY_DEFAULT_GATEWAY),
        proxy = values.proxy,
      )
      promise.resolve(null)
    } catch (e: Throwable) {
      promise.reject(e)
    }
  }
}

fun fromTheta(value: ConnectedWifiInfo): WritableMap {
  val result = Arguments.createMap()
  value.wifiInfo?.let {
    result.putMap(KEY_WIFI_INFO, fromTheta(it))
  }
  value.ethernet?.let {
    result.putMap(KEY_ETHERNET, fromTheta(it))
  }
  value.lte?.let {
    result.putMap(KEY_LTE, fromTheta(it))
  }
  return result
}

fun fromTheta(value: ConnectedInfo): WritableMap {
  val result = Arguments.createMap()
  result.putString(KEY_SSID, value.ssid)
  result.putBoolean(KEY_IS_CONNECTED, value.isConnected)
  result.putBoolean(KEY_IS_INTERNET_ACCESSIBLE, value.isInternetAccessible)
  return result
}

fun fromTheta(ssid: String?, ssidList: List<String>?): WritableMap {
  val result = Arguments.createMap()
  ssid?.let {
    result.putString(KEY_SSID, it)
  }
  ssidList?.let {
    val list = Arguments.fromList(it)
    result.putArray(KEY_SSID_LIST, list)
  }
  return result
}

data class SetAccessPointParams(
  val ssid: String,
  val ssidStealth: Boolean,
  val security: WifiSecurityMode,
  val password: String,
  val connectionPriority: Int,
  val ipAddress: String? = null,
  val subnetMask: String? = null,
  val defaultGateway: String? = null,
  val proxy: Proxy? = null,
)

fun toSetAccessPointParams(objects: ReadableMap): SetAccessPointParams {
  val ssid = objects.getString(KEY_SSID) ?: throw IllegalArgumentException(KEY_SSID)
  val ssidStealth = if (objects.hasKey(KEY_SSID_STEALTH)) {
    objects.getBoolean(KEY_SSID_STEALTH)
  } else {
    throw IllegalArgumentException(KEY_SSID_STEALTH)
  }
  val security =
    objects.getString(WifiSecurityMode.keyName)?.let { value ->
      WifiSecurityMode.entries.find {
        it.name == value
      }
    }
      ?: throw IllegalArgumentException(WifiSecurityMode.keyName)
  val password = objects.getString(KEY_PASSWORD) ?: throw IllegalArgumentException(KEY_PASSWORD)
  val connectionPriority = if (objects.hasKey(KEY_CONNECTION_PRIORITY)) {
    objects.getInt(KEY_CONNECTION_PRIORITY)
  } else {
    throw IllegalArgumentException(KEY_CONNECTION_PRIORITY)
  }
  val ipAddress = objects.getString(KEY_IP_ADDRESS)
  val subnetMask = objects.getString(KEY_SUBNET_MASK)
  val defaultGateway = objects.getString(KEY_DEFAULT_GATEWAY)
  val proxy = objects.getMap(Proxy.keyName)?.let { toProxy(it) }

  return SetAccessPointParams(
    ssid = ssid,
    ssidStealth = ssidStealth,
    security = security,
    password = password,
    connectionPriority = connectionPriority,
    ipAddress = ipAddress,
    subnetMask = subnetMask,
    defaultGateway = defaultGateway,
    proxy = proxy,
  )
}

File: react-native/ios/WlanControlCommandV2Service.swift
//
//  WlanControlCommandV2Service.swift
//  ThetaBleClientReactNative
//
//  Created on 2025/01/22.
//

import THETABleClient

let KEY_WIFI_INFO = "wifiInfo"
let KEY_ETHERNET = "ethernet"
let KEY_LTE = "lte"
let KEY_IS_CONNECTED = "isConnected"
let KEY_IS_INTERNET_ACCESSIBLE = "isInternetAccessible"
let KEY_SSID_LIST = "ssidList"
let KEY_SSID_STEALTH = "ssidStealth"
let KEY_CONNECTION_PRIORITY = "connectionPriority"
let KEY_IP_ADDRESS = "ipAddress"
let KEY_SUBNET_MASK = "subnetMask"
let KEY_DEFAULT_GATEWAY = "defaultGateway"

enum WlanControlCommandV2Service {
    static func setNetworkType(id: Int,
                               value: String,
                               resolve: @escaping RCTPromiseResolveBlock,
                               reject: @escaping RCTPromiseRejectBlock)
    {
        guard let device = ThetaBleClientReactNative.deviceList[id] else {
            reject(ERROR_TITLE, ERROR_MESSAGE_DEVICE_NOT_FOUND, nil)
            return
        }
        guard let service = device.wlanControlCommandV2 else {
            reject(ERROR_TITLE, ERROR_MESSAGE_UNSUPPORTED_SERVICE, nil)
            return
        }

        Task {
            do {
                let enumValue = getEnumValue(values: NetworkType.values(), name: value)
                guard let enumValue else {
                    reject(ERROR_TITLE, "Network type not found. \(value)", nil)
                    return
                }
                try await service.setNetworkType(networkType: enumValue)
                resolve(nil)
            } catch {
                reject(ERROR_TITLE, error.localizedDescription, error)
            }
        }
    }

    static func setNetworkTypeNotify(id: Int,
                                     enable: Bool,
                                     sendEvent: @escaping ([String: Any]) -> Void,
                                     resolve: @escaping () -> Void,
                                     reject: @escaping RCTPromiseRejectBlock)
    {
        guard let device = ThetaBleClientReactNative.deviceList[id] else {
            reject(ERROR_TITLE, ERROR_MESSAGE_DEVICE_NOT_FOUND, nil)
            return
        }
        guard let service = device.wlanControlCommandV2 else {
            reject(ERROR_TITLE, ERROR_MESSAGE_UNSUPPORTED_SERVICE, nil)
            return
        }

        do {
            if enable {
                try service.setNetworkTypeNotify { value, error in
                    sendEvent(
                        toNetworkTypeNotify(
                            deviceId: id,
                            value: value,
                            error: error
                        )
                    )
                }
            } else {
                try service.setNetworkTypeNotify()
            }
            resolve()
        } catch {
            reject(ERROR_TITLE, error.localizedDescription, error)
        }
    }

    static func getConnectedWifiInfo(id: Int,
                                     resolve: @escaping RCTPromiseResolveBlock,
                                     reject: @escaping RCTPromiseRejectBlock)
    {
        guard let device = ThetaBleClientReactNative.deviceList[id] else {
            reject(ERROR_TITLE, ERROR_MESSAGE_DEVICE_NOT_FOUND, nil)
            return
        }
        guard let service = device.wlanControlCommandV2 else {
            reject(ERROR_TITLE, ERROR_MESSAGE_UNSUPPORTED_SERVICE, nil)
            return
        }

        Task {
            do {
                let value = try await service.getConnectedWifiInfo()
                resolve(fromTheta(connectedWifiInfo: value))
            } catch {
                reject(ERROR_TITLE, error.localizedDescription, error)
            }
        }
    }

    static func setConnectedWifiInfoNotify(id: Int,
                                           enable: Bool,
                                           sendEvent: @escaping ([String: Any]) -> Void,
                                           resolve: @escaping () -> Void,
                                           reject: @escaping RCTPromiseRejectBlock)
    {
        guard let device = ThetaBleClientReactNative.deviceList[id] else {
            reject(ERROR_TITLE, ERROR_MESSAGE_DEVICE_NOT_FOUND, nil)
            return
        }
        guard let service = device.wlanControlCommandV2 else {
            reject(ERROR_TITLE, ERROR_MESSAGE_UNSUPPORTED_SERVICE, nil)
            return
        }

        do {
            if enable {
                try service.setConnectedWifiInfoNotify { value, error in
                    sendEvent(
                        toConnectedWifiInfoNotify(
                            deviceId: id,
                            value: value,
                            error: error
                        )
                    )
                }
            } else {
                try service.setConnectedWifiInfoNotify()
            }
            resolve()
        } catch {
            reject(ERROR_TITLE, error.localizedDescription, error)
        }
    }

    static func scanSsidStart(id: Int,
                              timeout: Int32,
                              sendEvent: @escaping ([String: Any]) -> Void,
                              resolve: @escaping RCTPromiseResolveBlock,
                              reject: @escaping RCTPromiseRejectBlock)
    {
        guard let device = ThetaBleClientReactNative.deviceList[id] else {
            reject(ERROR_TITLE, ERROR_MESSAGE_DEVICE_NOT_FOUND, nil)
            return
        }
        guard let service = device.wlanControlCommandV2 else {
            reject(ERROR_TITLE, ERROR_MESSAGE_UNSUPPORTED_SERVICE, nil)
            return
        }

        Task {
            do {
                class Callback: WlanControlCommandV2ScanCallback {
                    let deviceId: Int
                    let sendEvent: ([String: Any]) -> Void
                    init(deviceId: Int, sendEvent: @escaping ([String: Any]) -> Void) {
                        self.deviceId = deviceId
                        self.sendEvent = sendEvent
                    }

                    func onCompleted(ssidList: [String]) {
                        sendEvent(
                            toNotify(
                                deviceId: deviceId,
                                characteristic: BleCharacteristic.notificationScannedSsid,
                                params: fromTheta(ssid: nil, ssidList: ssidList),
                                error: nil
                            )
                        )
                    }

                    func onNotify(ssid: String) {
                        sendEvent(
                            toNotify(
                                deviceId: deviceId,
                                characteristic: BleCharacteristic.notificationScannedSsid,
                                params: fromTheta(ssid: ssid, ssidList: nil),
                                error: nil
                            )
                        )
                    }
                }
                try await service.scanSsidStart(timeout: timeout, callback: Callback(deviceId: id, sendEvent: sendEvent))
                resolve(nil)
            } catch {
                reject(ERROR_TITLE, error.localizedDescription, error)
            }
        }
    }

    static func scanSsidStop(id: Int,
                             resolve: @escaping RCTPromiseResolveBlock,
                             reject: @escaping RCTPromiseRejectBlock)
    {
        guard let device = ThetaBleClientReactNative.deviceList[id] else {
            reject(ERROR_TITLE, ERROR_MESSAGE_DEVICE_NOT_FOUND, nil)
            return
        }
        guard let service = device.wlanControlCommandV2 else {
            reject(ERROR_TITLE, ERROR_MESSAGE_UNSUPPORTED_SERVICE, nil)
            return
        }

        Task {
            do {
                try await service.scanSsidStop()
                resolve(nil)
            } catch {
                reject(ERROR_TITLE, error.localizedDescription, error)
            }
        }
    }

    static func setAccessPointDynamically(id: Int,
                                          params: [String: Any],
                                          resolve: @escaping RCTPromiseResolveBlock,
                                          reject: @escaping RCTPromiseRejectBlock)
    {
        guard let device = ThetaBleClientReactNative.deviceList[id] else {
            reject(ERROR_TITLE, ERROR_MESSAGE_DEVICE_NOT_FOUND, nil)
            return
        }
        guard let service = device.wlanControlCommandV2 else {
            reject(ERROR_TITLE, ERROR_MESSAGE_UNSUPPORTED_SERVICE, nil)
            return
        }

        Task {
            do {
                let base = try toSetAccessPointBasicParams(params: params)
                try await service.setAccessPointDynamically(
                    ssid: base.ssid,
                    ssidStealth: base.ssidStealth,
                    security: base.security,
                    password: base.password,
                    connectionPriority: base.connectionPriority,
                    proxy: base.proxy
                )
                resolve(nil)
            } catch {
                reject(ERROR_TITLE, error.localizedDescription, error)
            }
        }
    }

    static func setAccessPointStatically(id: Int,
                                         params: [String: Any],
                                         resolve: @escaping RCTPromiseResolveBlock,
                                         reject: @escaping RCTPromiseRejectBlock)
    {
        guard let device = ThetaBleClientReactNative.deviceList[id] else {
            reject(ERROR_TITLE, ERROR_MESSAGE_DEVICE_NOT_FOUND, nil)
            return
        }
        guard let service = device.wlanControlCommandV2 else {
            reject(ERROR_TITLE, ERROR_MESSAGE_UNSUPPORTED_SERVICE, nil)
            return
        }

        Task {
            do {
                let base = try toSetAccessPointBasicParams(params: params)
                let statically = try toSetAccessPointStaticallyParams(params: params)
                try await service.setAccessPointStatically(
                    ssid: base.ssid,
                    ssidStealth: base.ssidStealth,
                    security: base.security,
                    password: base.password,
                    connectionPriority: base.connectionPriority,
                    ipAddress: statically.ipAddress,
                    subnetMask: statically.subnetMask,
                    defaultGateway: statically.defaultGateway,
                    proxy: base.proxy
                )
                resolve(nil)
            } catch {
                reject(ERROR_TITLE, error.localizedDescription, error)
            }
        }
    }

    static func getWlanPasswordState(id: Int,
                                     resolve: @escaping RCTPromiseResolveBlock,
                                     reject: @escaping RCTPromiseRejectBlock)
    {
        guard let device = ThetaBleClientReactNative.deviceList[id] else {
            reject(ERROR_TITLE, ERROR_MESSAGE_DEVICE_NOT_FOUND, nil)
            return
        }
        guard let service = device.wlanControlCommandV2 else {
            reject(ERROR_TITLE, ERROR_MESSAGE_UNSUPPORTED_SERVICE, nil)
            return
        }

        Task {
            do {
                let value = try await service.getWlanPasswordState()
                resolve(value.name)
            } catch {
                reject(ERROR_TITLE, error.localizedDescription, error)
            }
        }
    }
}

func toNetworkTypeNotify(
    deviceId: Int,
    value: NetworkType?,
    error: KotlinThrowable?
) -> [String: Any] {
    let paramsInfo = (value != nil) ? [NetworkType.companion.keyName: value?.name] : nil
    let errorInfo = toNotifyError(error: error)
    return toNotify(deviceId: deviceId,
                    characteristic: BleCharacteristic.writeSetNetworkType,
                    params: paramsInfo,
                    error: errorInfo)
}

func fromTheta(connectedInfo: ConnectedInfo) -> [String: Any?] {
    let result: [String: Any?] = [
        KEY_SSID: connectedInfo.ssid,
        KEY_IS_CONNECTED: connectedInfo.isConnected,
        KEY_IS_INTERNET_ACCESSIBLE: connectedInfo.isInternetAccessible,
    ]
    return result
}

func fromTheta(connectedWifiInfo: ConnectedWifiInfo) -> [String: Any?] {
    var result: [String: Any?] = [:]
    if let value = connectedWifiInfo.wifiInfo {
        result[KEY_WIFI_INFO] = fromTheta(connectedInfo: value)
    }
    if let value = connectedWifiInfo.ethernet {
        result[KEY_ETHERNET] = fromTheta(connectedInfo: value)
    }
    if let value = connectedWifiInfo.lte {
        result[KEY_LTE] = fromTheta(connectedInfo: value)
    }
    return result
}

func toConnectedWifiInfoNotify(
    deviceId: Int,
    value: ConnectedWifiInfo?,
    error: KotlinThrowable?
) -> [String: Any] {
    var paramsInfo: [String: Any?]?
    if let value {
        paramsInfo = fromTheta(connectedWifiInfo: value)
    }
    let errorInfo = toNotifyError(error: error)
    return toNotify(deviceId: deviceId,
                    characteristic: BleCharacteristic.notificationConnectedWifiInfo,
                    params: paramsInfo,
                    error: errorInfo)
}

func fromTheta(ssid: String?, ssidList: [String]?) -> [String: Any?] {
    var result: [String: Any?] = [:]
    if let ssid {
        result[KEY_SSID] = ssid
    }
    if let ssidList {
        result[KEY_SSID_LIST] = ssidList
    }
    return result
}

struct SetAccessPointBaseParams {
    let ssid: String
    let ssidStealth: Bool
    let security: WifiSecurityMode
    let password: String
    let connectionPriority: Int32
    let proxy: Proxy?
}

struct SetAccessPointStaticallyParams {
    let ipAddress: String
    let subnetMask: String
    let defaultGateway: String
}

func toSetAccessPointBasicParams(params: [String: Any?]) throws -> SetAccessPointBaseParams {
    guard let ssid = params[KEY_SSID] as? String else {
        throw ThetaClientError.invalidArgument(KEY_SSID)
    }
    let ssidStealth = params[KEY_SSID_STEALTH] as? Bool ?? false
    let security = getEnumValue(
        values: WifiSecurityMode.values(),
        name: params[WifiSecurityMode.companion.keyName] as? String ?? WifiSecurityMode.none.name
    ) ?? WifiSecurityMode.none
    let password = params[KEY_PASSWORD] as? String ?? ""
    let connectionPriority = Int32(params[KEY_CONNECTION_PRIORITY] as? Int ?? 1)
    let ipAddress = params[KEY_IP_ADDRESS] as? String
    let subnetMask = params[KEY_SUBNET_MASK] as? String
    let defaultGateway = params[KEY_DEFAULT_GATEWAY] as? String
    let proxy = params[Proxy.companion.keyName]
    let proxyParam: Proxy? = {
        if let proxy = proxy as? [String: Any] {
            return toProxy(params: proxy)
        }
        return nil
    }()

    return SetAccessPointBaseParams(
        ssid: ssid,
        ssidStealth: ssidStealth,
        security: security,
        password: password,
        connectionPriority: connectionPriority,
        proxy: proxyParam
    )
}

func toSetAccessPointStaticallyParams(params: [String: Any?]) throws -> SetAccessPointStaticallyParams {
    guard let ipAddress = params[KEY_IP_ADDRESS] as? String else {
        throw ThetaClientError.invalidArgument(KEY_IP_ADDRESS)
    }
    guard let subnetMask = params[KEY_SUBNET_MASK] as? String else {
        throw ThetaClientError.invalidArgument(KEY_SUBNET_MASK)
    }
    guard let defaultGateway = params[KEY_DEFAULT_GATEWAY] as? String else {
        throw ThetaClientError.invalidArgument(KEY_DEFAULT_GATEWAY)
    }

    return SetAccessPointStaticallyParams(
        ipAddress: ipAddress,
        subnetMask: subnetMask,
        defaultGateway: defaultGateway
    )
}
